### PR TITLE
[RISCV] Inhibit DAG folding shl through zext.w pattern with zba

### PR DESCRIPTION
**Pull request by preames**
If we allow the fold, the zext.w pattern becomes an and by shifted 32 bit mask.  In practice, we can't undo this during ISEL resulting in worse code in some cases. There is a cost to inhibiting the generic transform -- we loose out on the possibility of folds enabled by pushing the shift earlier.

### CODE REVIEW

**Comment by topperc** [comment]
> In practice, we can't undo this during ISEL resulting in worse code in some cases.

Why can't we undo it? Does it move too far away?

**Comment by preames** [comment]
> > In practice, we can't undo this during ISEL resulting in worse code in some cases.
> 
> Why can't we undo it? Does it move too far away?

Sorry, I hadn't meant this as "we can't" and more "we don't".  We could write a non-trivial pattern match here, but we up having to match the whole (and (shl X, C), ShiftedMask) expression and then we have to check that C and the shift are compatible.  I looked at doing that, but it didn't seem particularly clean in tablegen.

File: llvm/test/CodeGen/RISCV/rv64zba.ll [hunk 1]
```diff
@@ -2891,8 +2890,7 @@ define i64 @srli_slliw(i64 %1) {
 ;
 ; RV64ZBA-LABEL: srli_slliw:
 ; RV64ZBA:       # %bb.0: # %entry
-; RV64ZBA-NEXT:    slli a0, a0, 2
-; RV64ZBA-NEXT:    srli a0, a0, 4
+; RV64ZBA-NEXT:    srli a0, a0, 2
 ; RV64ZBA-NEXT:    slli.uw a0, a0, 4
 ; RV64ZBA-NEXT:    ret
 entry:
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rv64zba.ll:2896]: Isn't this IR non-canonical per InstCombine

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/rv64zba.ll:2896]: The original test case is canonical: https://godbolt.org/z/ee3YPfY4s
```
define ptr @test(ptr %0, i64 %1) {
entry:
  %2 = lshr exact i64 %1, 2
  %3 = and i64 %2, 4294967295
  %4 = getelementptr inbounds i8, ptr %0, i64 600
  %5 = getelementptr [80 x i8], ptr %4, i64 %3
  ret ptr %5
}
```

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rv64zba.ll:2896]: Yes with a GEP its canonical, but with a shl it isn't.

Comment (preames) [reply llvm/test/CodeGen/RISCV/rv64zba.ll:2896]: Yeah, this was excessive reduction apparently.  

The fact instcombine prefers the opposite form does hint that we should maybe (also?) do the late match.  I was really hoping not to have to write that code...

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rv64zba.ll:2896]: I just posted it. We had the majority of the code already. Is there a 3 shift version of this we should do without Zba?

**Comment by preames** [comment]
Abandon in favor of Craig's alternative.

**Pull request closed without merging**